torchvision/datasets/widerface.py

```diff
@@ -0,0 +1,231 @@
+
+    def extra_repr(self) -> str:
+        lines = ["Target type: {target_type}", "Split: {split}"]
+        return '\n'.join(lines).format(**self.__dict__)
+
+    def parse_train_val_annotations_file(self) -> None:
+        filename = "wider_face_train_bbx_gt.txt" if self.split == "train" else "wider_face_val_bbx_gt.txt"
+        filepath = os.path.join(self.root, "wider_face_split", filename)
+
+        f = open(filepath, "r")
+        lines = f.readlines()
+
+        file_name_line, num_boxes_line, box_annotation_line = True, False, False
+        num_boxes, box_counter = 0, 0
+        labels = []
+        for line in lines:
+            line = line.rstrip()
+            if file_name_line:
+                img_path = os.path.join(self.root, "WIDER_" + self.split, "images", line)
+                img_path = abspath(expanduser(img_path))
+                file_name_line = False
+                num_boxes_line = True
+            elif num_boxes_line:
+                num_boxes = int(line)
+                num_boxes_line = False
+                box_annotation_line = True
+            elif box_annotation_line:
+                box_counter += 1
+                line_split = line.split(" ")
+                line_values = [int(x) for x in line_split]
+                labels.append(line_values)
+                if box_counter >= num_boxes:
+                    box_annotation_line = False
+                    file_name_line = True
+                    self.img_info.append({
+                        "img_path": img_path,
+                        "raw": torch.tensor(labels),
+                    })
+                    box_counter = 0
+                    labels.clear()
+            else:
+                raise RuntimeError("Error parsing annotation file {}".format(filepath))
+        f.close()
+
+    def parse_test_annotations_file(self) -> None:
+        filepath = os.path.join(self.root, "wider_face_split", "wider_face_test_filelist.txt")
+        filepath = abspath(expanduser(filepath))
+        f = open(filepath, "r")
+        lines = f.readlines()
+        for line in lines:
+            line = line.rstrip()
+            img_path = os.path.join(self.root, "WIDER_test", "images", line)
+            img_path = abspath(expanduser(img_path))
+            self.img_info.append({"img_path": img_path})
+        f.close()
+
+    def _check_integrity(self) -> bool:
+        # Allow original archive to be deleted (zip). Only need the extracted images
+        all_files = self.file_list.copy()
+        all_files.append(self.annotations_file)
+        for (_, md5, filename) in all_files:
+            file, ext = os.path.splitext(filename)
+            extracted_dir = os.path.join(self.root, file)
+            if os.path.exists(extracted_dir):
+                continue
+            filepath = os.path.join(self.root, filename)
+            if not check_integrity(filepath, md5):
+                return False
+            extract_archive(filepath)
```

Comment:
Unsure about this. Why would `check_integrity` extract the archive? Shouldn't this at least be in download?

Reply:
This covers the use case where someone has manually downloaded the zipped files to a filesystem but forgot to or has not yet unzipped the files. In this scenario, I do not expect the user to call `WIDERFace` with `download=True`...and we must still check and decompress the files if their respective folders do not exist.

Reply:
> In this scenario, I do not expect the user to call `WIDERFace` with `download=True`

That depends on the interpretation of `download=True`. For me (and this is how we handled this in other datasets) `download=True` means "please download **and** prepare the dataset so that I can use it". We could split this into `download` and `prepare` to be even more explicit. 

Still, IMO `check_integrity` is not the right place to put this. From the name I only expect it to check if everything is alright and not process the data in any way.

Reply:
@pmeier I've moved the dataset preparation logic so `download` downloads __and__ extracts the dataset files. `check_integrity(...)` behaves the way you describe now.

Unit tests have been added as well. I had to disable the tests on windows, otherwise error messages were raised regarding file permission errors when removing the dummy dataset instance. Whatever the issue is, it's outside the scope of the test based on what I can tell...I suspect this is the same reason why the cityscapes dataset tests are disabled on windows.